python/tests/e2e/conftest.py

```diff
@@ -51,22 +62,41 @@ class VCRConfig(TypedDict):
     - 'headers': Request headers
     """
 
-    filter_headers: list[str]
-    """Headers to filter out from recordings for security/privacy.
+    before_record_request: Callable[[Any], Any]
+    """Callback to sanitize requests before saving to cassette.
 
-    These headers will be removed from both recorded cassettes and
-    when matching requests during playback. Commonly used for:
-    - Authentication tokens
-    - API keys
-    - Organization identifiers
+    This function is called AFTER the real HTTP request is sent (with valid auth),
+    but BEFORE it's written to the cassette file. Use this to sanitize sensitive
+    headers without affecting the actual HTTP requests.
     """
 
-    filter_post_data_parameters: list[str]
-    """POST data parameters to filter out from recordings.
 
-    Similar to filter_headers but for form data and request body parameters.
-    Useful for removing sensitive data from request bodies.
+def sanitize_request(request: Any) -> Any:  # noqa: ANN401
+    """Sanitize sensitive headers in VCR request before recording to cassette.
+
+    This hook is called AFTER the real HTTP request is sent (with valid auth),
+    but BEFORE it's written to the cassette file. We deep copy the request
+    and replace sensitive headers with placeholders.
+
+    Args:
+        request: VCR request object to sanitize (Any type since VCR doesn't
+            provide typed request objects)
+
+    Returns:
+        Sanitized copy of the request safe for cassette storage
     """
+    request = deepcopy(request)
+
+    for header in SENSITIVE_HEADERS:
```

Comment:
nit: we could remove a loop and make it more efficient via:

```py
headers_to_filter = set(header.lower() for header in SENSITIVE_HEADERS)
for req_header in request.headers.keys():
  if req_header.lower() in headers_to_filter:
    ...
```

I'm going to treat this nit as optional because this code is test-only, loops are small fixed size constants